SU2_CFD/src/solvers/CScalarSolver.cpp

```diff
@@ -0,0 +1,958 @@
+  SU2_OMP_FOR_(schedule(static,omp_chunk_size) SU2_NOWAIT)
+  for (unsigned long iPoint = 0; iPoint < nPointDomain; iPoint++) {
+
+    /// TODO: This could be the SetTime_Step of this solver.
+    su2double dt = nodes->GetLocalCFL(iPoint) / flowNodes->GetLocalCFL(iPoint) * flowNodes->GetDelta_Time(iPoint);
+    nodes->SetDelta_Time(iPoint, dt);
+
+    /*--- Modify matrix diagonal to improve diagonal dominance. ---*/
+
+    if (dt != 0.0) {
+      su2double Vol = geometry->nodes->GetVolume(iPoint) + geometry->nodes->GetPeriodicVolume(iPoint);
+      Jacobian.AddVal2Diag(iPoint, Vol / dt);
+    }
+    else {
+      Jacobian.SetVal2Diag(iPoint, 1.0);
+      LinSysRes.SetBlock_Zero(iPoint);
+    }
+
+    /*--- Right hand side of the system (-Residual) and initial guess (x = 0) ---*/
+
+    for (unsigned short iVar = 0; iVar < nVar; iVar++) {
+      unsigned long total_index = iPoint*nVar + iVar;
+      LinSysRes[total_index] = -LinSysRes[total_index];
+      LinSysSol[total_index] = 0.0;
+
+      su2double Res = fabs(LinSysRes[total_index]);
+      resRMS[iVar] += Res*Res;
+      if (Res > resMax[iVar]) {
+        resMax[iVar] = Res;
+        idxMax[iVar] = iPoint;
+        coordMax[iVar] = geometry->nodes->GetCoord(iPoint);
+      }
+    }
+  }
+  END_SU2_OMP_FOR
+  SU2_OMP_CRITICAL
+  for (unsigned short iVar = 0; iVar < nVar; iVar++) {
+    Residual_RMS[iVar] += resRMS[iVar];
+    AddRes_Max(iVar, resMax[iVar], geometry->nodes->GetGlobalIndex(idxMax[iVar]), coordMax[iVar]);
+  }
+  END_SU2_OMP_CRITICAL
+  SU2_OMP_BARRIER
+
+  /*--- Compute the root mean square residual ---*/
+  SetResidual_RMS(geometry, config);
+}
+
+void CScalarSolver::CompleteImplicitIteration(CGeometry *geometry, CSolver **solver_container, CConfig *config) {
+
+  const bool compressible = (config->GetKind_Regime() == ENUM_REGIME::COMPRESSIBLE);
+
+  const auto flowNodes = solver_container[FLOW_SOL]->GetNodes();
+
+  ComputeUnderRelaxationFactor(config);
+
+  /*--- Update solution (system written in terms of increments) ---*/
+
+  if (!adjoint) {
+
+    /*--- Update the turbulent solution. Only SST variants are clipped. ---*/
+
+    switch (config->GetKind_Turb_Model()) {
+
+      case SA: case SA_E: case SA_COMP: case SA_E_COMP: case SA_NEG:
+
+        SU2_OMP_FOR_STAT(omp_chunk_size)
+        for (unsigned long iPoint = 0; iPoint < nPointDomain; iPoint++) {
+          nodes->AddSolution(iPoint, 0, nodes->GetUnderRelaxation(iPoint)*LinSysSol[iPoint]);
+        }
+        END_SU2_OMP_FOR
+        break;
+
+      case SST: case SST_SUST:
+
+        SU2_OMP_FOR_STAT(omp_chunk_size)
+        for (unsigned long iPoint = 0; iPoint < nPointDomain; iPoint++) {
+
+          su2double density = flowNodes->GetDensity(iPoint);
+          su2double density_old = density;
+
+          if (compressible)
+            density_old = flowNodes->GetSolution_Old(iPoint,0);
+
+          for (unsigned short iVar = 0; iVar < nVar; iVar++) {
+            nodes->AddConservativeSolution(iPoint, iVar,
```

Comment:
Ah... yet another variant, well we can always override those very special cases. It is probably better to have one or two exceptions than to shoe-horn every solver into an interface that will be inefficient in 90% of cases (every extra variable is a little more memory for AD for example).

Reply:
I added a `bool Conservative` to Cscalarsolver as class data and used the more general `AddConservativeSolution` for both, the clipping limits are default set to su2doule::min/max in case a solver does not use this (e.g. SA)